### PR TITLE
Add pytest-regressions to replace port tests

### LINKED ISSUE #42: Replace existing port tests with automated regression tests
Currently, we test our estimator accuracy against manually generated results from `yaImpute` and `pynnmap`. Once we have all major functionality implemented and are confident in our results, we can switch to regression testing to ensure no errors are introduced. This was briefly discussed in #40:

> Eventually, I suppose we will reach a point where we're confident everything is working and could theoretically test against previous versions of sknnr ... [pytest-regressions](https://github.com/ESSS/pytest-regressions) and [syrupy](https://github.com/tophat/syrupy) both look like interesting tools that might solve this problem for us in the future by testing against automatically generated results.

We'll need to do some experimenting to find the right tool for our use case, but the advantage should be a massive simplification of our testing system (e.g. the removal of `KNNTestDataset`) and the ability to easily test against a wider range of parameters (e.g. values of `n_neighbors` and `n_components`).

### PR DESCRIPTION
**Pull request by grovduck**
Closes #42 

This PR add `pytest-regressions` as a dependency of the hatch `test` environment and creates regression data via `ndarrays_regression`.  The data that is stored would replace the data that we captured from `yaImpute` as part of the porting process.  Now that most estimators have been ported and verified from `yaImpute`, using automated regression testing is a better path forward.

To implement the change, we followed this process:
#### Setup
- Verified that all tests pass on `main` branch with old `yaImpute` CSV files
- Checked out this new branch (`pytest-regressions`)
- Made `pytest-regressions` a dependency of hatch `test` environment
#### Editing tests and regression data
- Copied `test_port.py::test_predict` to `test_port.py::test_predict_regressions`
- Edited `test_predict_regressions` to create the regression files
- Ran `test_predict_regressions` - all failed on first round, but the tests created a new folder called `test_port` with all combinations of input parameters as `.npz` files
- Ran `test_predict_regressions` again - all pass
- Copied `test_port.py::test_kneighbors` to `test_port.py::test_kneighbors_regressions`
- Ran `test_kneighbors_regressions` - all fail on first round but succeed on second round as before
- Added `reference` as a pytest parameter and reran - all pass
#### Verified against yaImpute test files
- Wrote script (temporarily committed as `verify_regression_data`) to verify that regression data matches yaimpute data
#### Consolidated score into predict test
- Incorporated independent score into `test_predict_regressions` so that `reference` parameterizations have an additional "score" key (in addition to the "pred" key).  The target files only have the "pred" key.
- Ran `--force-regen` and `-k reference` on `test_predict_regressions` to regenerate the `reference` based regression files - reran - all tests failed but regenerated files.  Ran again without `--force-regen` and it worked correctly

There are a few edits and decisions remaining to finish this PR:
- Remove all `yaImpute` test files once team can verify that output matches
- Modify `load_moscow_stjoes_results` to only load data and create the train-test split.  We don't need to bring in any of the "check" data as that is now the role of the regression data (which nicely uses the pytest parameters to name output files).  This would also include modifying (and potentially renaming) `KNNTestDataset` to exclude test data members.
- Decide if we want to add regression tests/data for the southwest Oregon data as well
- It looks like there is [some flexibility](https://pytest-regressions.readthedocs.io/en/latest/overview.html#data-directory-path) on naming output directories and files.  Right now, the regression `.npz` files are put into a directory named after the test.

### CODE REVIEW

**Comment by grovduck** [comment]
@aazuspan, as I was working on the RFNN branch yesterday, I realized that the default value I had for one of the regression parameters was not the same default as `scikit-learn`.  Setting it to the correct default meant that all port tests failed.  So, I figured this was as good a time as any to get the regression tests implemented.

Right now, this is a draft (e.g. I've left the old port tests and data as-is), but I'm hoping that you can test two things for me before I proceed with other changes.

- Verify that the new tests pass for you.   Because the new regression data will be in place for you, these tests should pass.
```
hatch run test:all tests\test_port.py::test_kneighbors_regressions
hatch run test:all tests\test_port.py::test_predict_regressions
```
- Run the script `verify_regression_data.py` at the `sknnr` root.  I didn't put a ton of time into it, so it's not beautiful, but is basically comparing that the array data from neighbors, distances, and predictions are all matching.  We had discussed this approach in the [related issue](https://github.com/lemma-osu/sknnr/issues/42#issuecomment-1718398310).  I'll remove this script once you've verified it's working for you.

Let me know if you see anything else right off or have any comments on the questions/issues I posed above.

**Comment by aazuspan** [comment]
This is exciting! The tests and script all passed on my end, and your logic for validating the regression test results seems sound.

> Decide if we want to add regression tests/data for the southwest Oregon data as well

I could see this if you think there's a reason that errors might arise here that don't with the Moscow data? Otherwise, it might not be worth the extra testing time.

> It looks like there is [some flexibility](https://pytest-regressions.readthedocs.io/en/latest/overview.html#data-directory-path) on naming output directories and files. Right now, the regression .npz files are put into a directory named after the test.

I don't have any strong feelings on this, so I'm okay with the defaults unless you have other thoughts.

Other than those issues, I think having regression tests opens up the possibility to parametrize over more options like "k" and "return_dataframe_index", but it probably makes sense to tackle that separately?

Let me know when you're ready for a more thorough review!

**Comment by grovduck** [comment]
Thanks for the feedback, @aazuspan!  A few comments and then I think this is ready for your review:

> I could see this if you think there's a reason that errors might arise here that don't with the Moscow data? Otherwise, it might not be worth the extra testing time.

I did a one-time check against SWO files (that I had locally) and they all passed.  Tests were a bit slow as you surmised and it added one combination level, so not really worth keeping around.

> I don't have any strong feelings on this, so I'm okay with the defaults unless you have other thoughts.

Sounds good.  I left this as-is (`test_port`).  I'm wondering if we want to now rename this to something like `test_regressions.py`, but then we might be losing the historical context of why we did this in the first place.  The meaning of "port" might get lost over time, though.  Any thoughts?

> Other than those issues, I think having regression tests opens up the possibility to parametrize over more options like "k" and "return_dataframe_index", but it probably makes sense to tackle that separately?

That's a great idea.  The lack of `k` in the output filenames is definitely noticeable and I agree that testing `return_dataframe_index` (with shorter aliases!) would be good as well.  I'm OK tackling it here or we could push to a new PR.

I ended up renaming `load_moscow_stjoes_results` to `load_moscow_stjoes_test_data` but kept the `KNNTestDataset` name as it still reflects the input data.  Obviously it is thinned way down.  The function `load_moscow_stjoes_test_data` now takes no parameters, so we might want to make it a pytest.fixture (and move to `conftest.py`)?

I'll go ahead and do the Copilot review to catch my errors, but look forward to your review as well.

**Review by aazuspan — CHANGES_REQUESTED** [review]
This looks great, @grovduck! Cleaning out all the cruft around the old port code is very satisfying, and will make the tests much more flexible and easy to expand.

> I'm wondering if we want to now rename this to something like test_regressions.py, but then we might be losing the historical context of why we did this in the first place. The meaning of "port" might get lost over time, though. Any thoughts?

These are both good points. I think I land on the side of renaming to `test_regressions.py` to make the current usage clear.

>  The lack of k in the output filenames is definitely noticeable and I agree that testing return_dataframe_index (with shorter aliases!) would be good as well. I'm OK tackling it here or we could push to a new PR.

Tackling here is good with me! Merging all the new test files in one go would be nice.

File: tests/test_port.py [hunk 1]
```diff
@@ -48,75 +45,46 @@ def get_default_hyperparams(estimator, **kwargs) -> dict:
...
 @pytest.mark.parametrize("reference", [True, False], ids=["reference", "target"])
-def test_predict(result, n_components, weighted, reference):
-    """Test that the ported estimators predict the correct values."""
-    method, estimator = result
-    dataset = load_moscow_stjoes_results(method=method, n_components=n_components)
+def test_predict(ndarrays_regression, estimator, n_components, weighted, reference):
+    """Test that the ported estimators predict and score the correct values."""
+    dataset = load_moscow_stjoes_test_data()
 
     weights = yaimpute_weights if weighted else None
```
[the remark this answers is not in the document]
Comment (grovduck) [reply tests/test_port.py]: Let me know if I hit the mark on what you were thinking here.  Happy to revise as necessary.

File: tests/datasets.py [hunk 2]
```diff
@@ -1,94 +1,33 @@
...
 from numpy.typing import NDArray
 from sklearn.model_selection import train_test_split
 
 from sknnr.datasets import load_moscow_stjoes
 
-TEST_DATA_MODULE = "tests.data"
-
 
 @dataclass
 class KNNTestDataset:
```
[the remark this answers is not in the document]
Comment (grovduck) [reply tests/datasets.py]: Agreed.  Renamed to `TrainTestDataset`.

File: tests/datasets.py [hunk 3]
```diff
@@ -1,94 +1,33 @@
...
-    trg_neighbors: NDArray[np.float64]
-    trg_predicted_weighted: NDArray[np.float64]
-    trg_predicted_unweighted: NDArray[np.float64]
 
 
-def load_moscow_stjoes_results(
-    method: str, k: int = 5, n_components: int | None = None
-) -> KNNTestDataset:
-    """Load the Moscow Mountain / St. Joe's dataset results for port testing.
+def load_moscow_stjoes_test_data() -> KNNTestDataset:
```
[the remark this answers is not in the document]
Comment (grovduck) [reply tests/datasets.py]: Great.  I've moved the dataclass `TrainTestDataset` and renamed the function `load_moscow_stjoes_test_data` to a pytest fixture called `moscow_stjoes_tests_data`.  Once this was done, `tests/datasets.py` could be deleted.

File: tests/test_port.py [hunk 4]
```diff
@@ -48,75 +45,46 @@ def get_default_hyperparams(estimator, **kwargs) -> dict:
 
 @pytest.mark.uncollect_if(func=estimator_does_not_support_n_components)
 @pytest.mark.parametrize(
-    "result", ESTIMATOR_RESULTS.items(), ids=ESTIMATOR_RESULTS.keys()
+    "estimator", ESTIMATOR_RESULTS.values(), ids=ESTIMATOR_RESULTS.keys()
```
[the remark this answers is not in the document]
Comment (grovduck) [reply tests/test_port.py]: Good idea.  I've done this.

**Comment by grovduck** [comment]
@aazuspan, thanks for the helpful review.  

> These are both good points. I think I land on the side of renaming to `test_regressions.py` to make the current usage clear.

I've renamed `test_port.py` to `test_regressions.py`.

> Tackling here is good with me! Merging all the new test files in one go would be nice.

Perfect.  `k` and `return_dataframe_index` are now parameterizations.  Right now, I'm only testing k=5, but we can expand if there is any reason to do so.  For `return_dataframe_index`, the output IDs are `ids` for `True` and `index` for `False`.

Let me know if anything else catches your eye!

File: tests/test_port.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aazuspan) [reply tests/test_port.py]: The docstring is a good addition, but it might be helpful to have a comment to keep the context of why we're continuing to test with that weighting function. I assume the reasoning is that it keeps a regression test with outputs equivalent to the port tests against `yaimpute`? If that's right, then maybe something along the lines of "This weighting generates predictions equivalent to yaImpute so we can continue to test correctness of ported estimators"?

**Review by aazuspan — APPROVED** [review]
Thanks for the changes - it all looks great to me!

File: tests/test_port.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (grovduck) [reply tests/test_port.py]: Excellent, thanks for this clarification.  I took your wording verbatim and added it as a comment rather than as a docstring as I think it is more important for us (rather than users) to track.
